src/server/injected/injectedScript.ts

```diff
@@ -267,35 +271,52 @@ export class InjectedScript {
-    if (!element || !element.isConnected)
-      return 'error:notconnected';
-    if (element.nodeName.toLowerCase() !== 'select')
-      return 'error:notselect';
-    const select = element as HTMLSelectElement;
-    const options = Array.from(select.options);
-    select.value = undefined as any;
-    for (let index = 0; index < options.length; index++) {
-      const option = options[index];
-      option.selected = optionsToSelect.some(optionToSelect => {
-        if (optionToSelect instanceof Node)
-          return option === optionToSelect;
-        let matches = true;
-        if (optionToSelect.value !== undefined)
-          matches = matches && optionToSelect.value === option.value;
-        if (optionToSelect.label !== undefined)
-          matches = matches && optionToSelect.label === option.label;
-        if (optionToSelect.index !== undefined)
-          matches = matches && optionToSelect.index === index;
-        return matches;
-      });
-      if (option.selected && !select.multiple)
-        break;
-    }
-    select.dispatchEvent(new Event('input', { 'bubbles': true }));
-    select.dispatchEvent(new Event('change', { 'bubbles': true }));
-    return options.filter(option => option.selected).map(option => option.value);
+  waitForOptionsAndSelect(node: Node, optionsToSelect: (Node | { value?: string, label?: string, index?: number })[]): InjectedScriptPoll<string[] | 'error:notconnected' | FatalDOMError> {
+    return this.pollRaf((progress, continuePolling) => {
+      const element = this.findLabelTarget(node as Element);
+      if (!element || !element.isConnected)
+        return 'error:notconnected';
+      if (element.nodeName.toLowerCase() !== 'select')
+        return 'error:notselect';
+      const select = element as HTMLSelectElement;
+      const options = Array.from(select.options);
+      const selectedOptions = [];
+      let remainingOptionsToSelect = optionsToSelect.slice();
+      for (let index = 0; index < options.length; index++) {
+        const option = options[index];
+        const filter = (optionToSelect: Node | { value?: string, label?: string, index?: number }) => {
+          if (optionToSelect instanceof Node)
+            return option === optionToSelect;
+          let matches = true;
+          if (optionToSelect.value !== undefined)
+            matches = matches && optionToSelect.value === option.value;
+          if (optionToSelect.label !== undefined)
+            matches = matches && optionToSelect.label === option.label;
+          if (optionToSelect.index !== undefined)
+            matches = matches && optionToSelect.index === index;
+          return matches;
+        };
+        if (!remainingOptionsToSelect.some(filter))
+          continue;
+        selectedOptions.push(option);
+        if (select.multiple) {
+          remainingOptionsToSelect = remainingOptionsToSelect.filter(o => !filter(o));
+        } else {
+          remainingOptionsToSelect = [];
```

Comment:
This logic seems strange. I guess we should throw when multiple options are passed for non-multiple select? Otherwise there is no way we select them all.

Reply:
From our docs: " If the <select> has the multiple attribute, all matching options are selected, otherwise only the first option matching one of the passed options is selected."